src/main/java/com/cloudbees/hudson/plugins/folder/AbstractFolderDescriptor.java

```diff
@@ -183,4 +189,53 @@ public boolean isLookAndFeelConfigurable(AbstractFolder<?> folder) {
+     * always. The obvious solution is that the {@link ComputedFolder} should mangle the supplied {@link Item#getName()}
+     * but the side-effect is that the URLs of the child items will now be mangled. Additionally the filename
+     * requirements can be rather onerous. Here is the most portable list of filename specification:
+     * <ul>
+     *     <li>Assume case insensitive</li>
+     *     <li>Assume no filename can be longer than 32 characters</li>
+     *     <li>Assume that only the characters {@code A-Za-z0-9_.-} are available</li>
+     *     <li>Assume that there are some special reserved names such as {@code .} and {@code ..}</li>
+     *     <li>Assume that there are some problematic names to avoid such as {@code AUX}, {@code CON}, {@code NUL}, etc.
+     *         See <a href="https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx">
+     *         Microsoft's page on "Naming Files, Paths, and Namespaces"</a>
+     *         (What's that you say, "Oh but we only run on Linux", perhaps but users may want to migrate from one OS
+     *         to an other by <strong>moving</strong> their {@code JENKINS_HOME} (or even parts of it) so if you are
+     *         mangling names, be sure to ensure that the mangled name is the same on all OSes
+     *     </li>
+     *     <li><a href="http://unicode.org/reports/tr15/">NFC vs NFD</a> may be a concern as different filesystems
+     *     apply different rules and normalization to the filenames. This is primarily a concern when migrating from
+     *     before having a {@link ChildNameGenerator} to having a {@link ChildNameGenerator} as the migration will
+     *     require inference of the unmangled name from the filesystem, which may or may not match the unmangeld
+     *     name from the source of the computation. Now POSIX does not specify how the filesystem is supposed to handle
+     *     encoding of filenames and there can be strange behaviours, e.g.
+     *     <a href="http://stackoverflow.com/a/32663908/1589700">{@link File#listFiles()} is rumoured to always return
+     *     NFC on OS-X</a>
+     *     </li>
+     * </ul>
+     * The {@link ChildNameGenerator} at least allows an {@link AbstractFolder} to apply an on-disk naming strategy
+     * that differs from the names used for the URLs.
+     * <p>
+     * <strong>IF YOU IMPLEMENT A {@link ChildNameGenerator} YOU MUST RETURN A SINGLETON FROM THIS METHOD</strong>
+     *
+     * @param <I> A wildcard parameter to assist type matching.
+     * @return a <strong>SINGLETON</strong> instance of {@link ChildNameGenerator} for all instanced of the concrete
+     * {@link AbstractFolder} class or {@code null} if no name mangling will be performed.
+     * @since 5.17
+     */
+    // TODO figure out how one could un-wind name mangling if one ever wanted to
+    @CheckForNull
+    public <I extends TopLevelItem> ChildNameGenerator<AbstractFolder<I>,I> childNameGenerator() {
+        return null;
```

Comment:
Is there not a sensible default implementation, at least for `ComputedFolder`s? I would not want a scary complex implementation to be copied among downstream plugins.